misc/multistream-select/src/protocol.rs

```diff
@@ -50,21 +50,72 @@ const MAX_PROTOCOL_LEN: usize = 140;
 
 /// The encoded form of a multistream-select 1.0.0 header message.
 const MSG_MULTISTREAM_1_0: &[u8] = b"/multistream/1.0.0\n";
-/// The encoded form of a multistream-select 2.0.0 header message.
-const MSG_MULTISTREAM_2_0: &[u8] = b"/multistream/2.0.0\n";
+/// The encoded form of a multistream-select 1.0.0 header message.
+const MSG_MULTISTREAM_1_0_LAZY: &[u8] = b"/multistream-lazy/1\n";
```

Comment:
Hmmm, I thought the idea was to keep `/multistream/1.0.0`, but I guess this works as well.

Reply:
Without a different protocol name, there can be no different `Version`s based on which both dialer and listener implement different behaviour (e.g. when to flush). I don't see how that could be done otherwise.

Reply:
From the client's point of view, dialing `V1` and `V1Lazy` is the same (except for the protocol name), no?

Reply:
> From the client's point of view, dialing `V1` and `V1Lazy` is the same (except for the protocol name), no?

No, for `V1` the dialer now never prematurely goes into the `Expecting` state, it flushes every protocol proposal and reads the reply. This was done previously in  `libp2p_core::upgrade::apply_outbound` anyway by waiting for negotiation to complete between upgrades, as I mentioned in the PR description, to avoid the edge-cases discussed in the other PRs. Now this behaviour is localised to `multistream-select` and subject to the chosen version.

Also, the listener needs to know whether the dialer wants the lazy variant or not, based on which it decides whether it needs to immediately flush its protocol response (addressing https://github.com/libp2p/rust-libp2p/issues/1241 as mentioned in the PR description). Without a different protocol name on the wire, there can be no parsing into different `Version`s.

Apart from that, even though `V1` and `V1Lazy` are identical as far as the protocol messages on the wire are concerned, important expectations on the granularity with which requests and responses are sent and received or buffered that can cause incompatibilities if not observed by both peers are part of the protocol and thus such differences deserve a different protocol name from that perspective as well, I think.